File: lib/codepipeline_stack.py
"""codepipeline_stack.py
This module defines the CodePipelineStack class, which creates the CodePipeline
for orchestrating the CI/CD workflow with build, staging, and production stages.
"""

from typing import Optional

import aws_cdk as cdk
from aws_cdk import aws_codebuild as codebuild
from aws_cdk import aws_codecommit as codecommit
from aws_cdk import aws_codepipeline as codepipeline
from aws_cdk import aws_codepipeline_actions as codepipeline_actions
from aws_cdk import aws_iam as iam
from aws_cdk import aws_s3 as s3
from constructs import Construct
class CodePipelineStackProps(cdk.StackProps):
  """
  CodePipelineStackProps defines the properties for the CodePipelineStack.
  
  Args:
    environment_suffix (Optional[str]): Environment suffix for resource naming
    codepipeline_role (iam.Role): IAM role for CodePipeline
    artifacts_bucket (s3.Bucket): S3 bucket for storing pipeline artifacts
    build_project (codebuild.Project): CodeBuild project for building code
    deploy_staging_project (codebuild.Project): CodeBuild project for staging deployment
    deploy_production_project (codebuild.Project): CodeBuild project for production deployment
    **kwargs: Additional keyword arguments passed to the base cdk.StackProps
  """
  
  def __init__(
    self,
    environment_suffix: Optional[str] = None,
    codepipeline_role: Optional[iam.Role] = None,
    artifacts_bucket: Optional[s3.Bucket] = None,
    build_project: Optional[codebuild.Project] = None,
    deploy_staging_project: Optional[codebuild.Project] = None,
    deploy_production_project: Optional[codebuild.Project] = None,
    **kwargs
  ):
    super().__init__(**kwargs)
    self.environment_suffix = environment_suffix
    self.codepipeline_role = codepipeline_role
    self.artifacts_bucket = artifacts_bucket
    self.build_project = build_project
    self.deploy_staging_project = deploy_staging_project
    self.deploy_production_project = deploy_production_project
class CodePipelineStack(cdk.NestedStack):
  """
  CodePipelineStack creates the CI/CD pipeline for automated deployments.
  
  This stack creates:
  - CodeCommit repository for source code
  - CodePipeline with source, build, staging, and production stages
  - Manual approval action between staging and production
  
  Args:
    scope (Construct): The parent construct
    construct_id (str): The unique identifier for this stack
    props (Optional[CodePipelineStackProps]): Optional properties for configuring the stack
    **kwargs: Additional keyword arguments passed to the CDK NestedStack
  """
  
  def __init__(
    self,
    scope: Construct,
    construct_id: str,
    props: Optional[CodePipelineStackProps] = None,
    **kwargs
  ):
    super().__init__(scope, construct_id, **kwargs)
    
    environment_suffix = props.environment_suffix if props else 'dev'
    codepipeline_role = props.codepipeline_role if props else None
    artifacts_bucket = props.artifacts_bucket if props else None
    build_project = props.build_project if props else None
    deploy_staging_project = props.deploy_staging_project if props else None
    deploy_production_project = props.deploy_production_project if props else None
    
    # Create CodeCommit repository
    self.repository = codecommit.Repository(
      self,
      "Repository",
      repository_name=f"ciapp-{environment_suffix}-repo",
      description=(f"Source code repository for CI/CD pipeline - "
                   f"{environment_suffix} environment")
    )
    
    # Define pipeline artifacts
    source_output = codepipeline.Artifact("SourceOutput")
    build_output = codepipeline.Artifact("BuildOutput")
    
    # Create the pipeline
    self.pipeline = codepipeline.Pipeline(
      self,
      "Pipeline",
      pipeline_name=f"ciapp-{environment_suffix}-pipeline",
      role=codepipeline_role,
      artifact_bucket=artifacts_bucket,
      stages=[
        # Source stage
        codepipeline.StageProps(
          stage_name="Source",
          actions=[
            codepipeline_actions.CodeCommitSourceAction(
              action_name="Source",
              repository=self.repository,
              branch="main",
              output=source_output,
              trigger=codepipeline_actions.CodeCommitTrigger.EVENTS
            )
          ]
        ),
        
        # Build stage
        codepipeline.StageProps(
          stage_name="Build",
          actions=[
            codepipeline_actions.CodeBuildAction(
              action_name="Build",
              project=build_project,
              input=source_output,
              outputs=[build_output]
            )
          ]
        ),
        
        # Deploy to Staging stage
        codepipeline.StageProps(
          stage_name="DeployStaging",
          actions=[
            codepipeline_actions.CodeBuildAction(
              action_name="DeployToStaging", 
              project=deploy_staging_project,
              input=build_output
            )
          ]
        ),
        
        # Manual approval stage
        codepipeline.StageProps(
          stage_name="ApproveProduction",
          actions=[
            codepipeline_actions.ManualApprovalAction(
              action_name="ApproveProductionDeployment",
              additional_information=("Please review the staging deployment "
                                      "and approve for production deployment.")
            )
          ]
        ),
        
        # Deploy to Production stage
        codepipeline.StageProps(
          stage_name="DeployProduction",
          actions=[
            codepipeline_actions.CodeBuildAction(
              action_name="DeployToProduction",
              project=deploy_production_project,
              input=build_output
            )
          ]
        )
      ]
    )
    
    # Output the repository clone URL
    cdk.CfnOutput(
      self,
      "RepositoryCloneUrl",
      value=self.repository.repository_clone_url_http,
      description="HTTP clone URL for the CodeCommit repository"
    )
    
    # Output the pipeline console URL
    cdk.CfnOutput(
      self,
      "PipelineUrl",
      value=(f"https://console.aws.amazon.com/codesuite/codepipeline/"
             f"pipelines/{self.pipeline.pipeline_name}/view"),
      description="AWS Console URL for the CodePipeline"
    )
    
    # Add tags to all resources
    cdk.Tags.of(self).add("Environment", environment_suffix)
    cdk.Tags.of(self).add("Component", "Pipeline")

File: lib/codepipeline-stack.ts
/**
 * codepipeline-stack.ts
 *
 * Defines AWS CodePipeline with 5 stages: Source, Build, Test, Security, Deploy.
 *
 * Features:
 * - CodeCommit source stage
 * - CodeBuild stages for build, test, security
 * - Manual approval before deployment
 * - Lambda invoke for blue-green deployment
 * - KMS encryption for artifacts
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface CodePipelineStackArgs {
  environmentSuffix: string;
  repositoryName: pulumi.Input<string>;
  artifactsBucket: pulumi.Input<string>;
  kmsKeyArn: pulumi.Input<string>;
  buildProjectName: pulumi.Input<string>;
  testProjectName: pulumi.Input<string>;
  securityProjectName: pulumi.Input<string>;
  deployFunctionName: pulumi.Input<string>;
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class CodePipelineStack extends pulumi.ComponentResource {
  public readonly pipeline: aws.codepipeline.Pipeline;
  public readonly pipelineRole: aws.iam.Role;

  constructor(
    name: string,
    args: CodePipelineStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:codepipeline:CodePipelineStack', name, args, opts);

    const {
      environmentSuffix,
      repositoryName,
      artifactsBucket,
      kmsKeyArn,
      buildProjectName,
      testProjectName,
      securityProjectName,
      deployFunctionName,
      tags,
    } = args;

    // Pipeline role
    this.pipelineRole = new aws.iam.Role(
      `codepipeline-role-${environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'codepipeline.amazonaws.com' },
              Action: 'sts:AssumeRole',
            },
          ],
        }),
        tags: {
          ...tags,
          Name: `codepipeline-role-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    new aws.iam.RolePolicy(
      `codepipeline-policy-${environmentSuffix}`,
      {
        role: this.pipelineRole.id,
        policy: pulumi
          .all([artifactsBucket, kmsKeyArn, deployFunctionName])
          .apply(([bucket, keyArn, funcName]) =>
            JSON.stringify({
              Version: '2012-10-17',
              Statement: [
                {
                  Effect: 'Allow',
                  Action: [
                    's3:GetObject',
                    's3:PutObject',
                    's3:GetBucketLocation',
                  ],
                  Resource: [
                    `arn:aws:s3:::${bucket}`,
                    `arn:aws:s3:::${bucket}/*`,
                  ],
                },
                {
                  Effect: 'Allow',
                  Action: ['kms:Decrypt', 'kms:Encrypt', 'kms:GenerateDataKey'],
                  Resource: keyArn,
                },
                {
                  Effect: 'Allow',
                  Action: [
                    'codecommit:GetBranch',
                    'codecommit:GetCommit',
                    'codecommit:UploadArchive',
                    'codecommit:GetUploadArchiveStatus',
                  ],
                  Resource: '*',
                },
                {
                  Effect: 'Allow',
                  Action: ['codebuild:BatchGetBuilds', 'codebuild:StartBuild'],
                  Resource: '*',
                },
                {
                  Effect: 'Allow',
                  Action: ['lambda:InvokeFunction'],
                  Resource: `arn:aws:lambda:*:*:function:${funcName}`,
                },
              ],
            })
          ),
      },
      { parent: this }
    );

    // Create Pipeline
    this.pipeline = new aws.codepipeline.Pipeline(
      `cicd-pipeline-${environmentSuffix}`,
      {
        name: `cicd-pipeline-${environmentSuffix}`,
        roleArn: this.pipelineRole.arn,
        artifactStores: [
          {
            location: artifactsBucket,
            type: 'S3',
            encryptionKey: {
              id: kmsKeyArn,
              type: 'KMS',
            },
          },
        ],
        stages: [
          // Stage 1: Source
          {
            name: 'Source',
            actions: [
              {
                name: 'SourceAction',
                category: 'Source',
                owner: 'AWS',
                provider: 'CodeCommit',
                version: '1',
                outputArtifacts: ['SourceOutput'],
                configuration: {
                  RepositoryName: repositoryName,
                  BranchName: 'main',
                  PollForSourceChanges: 'false', // Use EventBridge instead
                },
              },
            ],
          },
          // Stage 2: Build
          {
            name: 'Build',
            actions: [
              {
                name: 'BuildAction',
                category: 'Build',
                owner: 'AWS',
                provider: 'CodeBuild',
                version: '1',
                inputArtifacts: ['SourceOutput'],
                outputArtifacts: ['BuildOutput'],
                configuration: {
                  ProjectName: buildProjectName,
                },
              },
            ],
          },
          // Stage 3: Test
          {
            name: 'Test',
            actions: [
              {
                name: 'TestAction',
                category: 'Test',
                owner: 'AWS',
                provider: 'CodeBuild',
                version: '1',
                inputArtifacts: ['BuildOutput'],
                outputArtifacts: ['TestOutput'],
                configuration: {
                  ProjectName: testProjectName,
                },
              },
            ],
          },
          // Stage 4: Security
          {
            name: 'Security',
            actions: [
              {
                name: 'SecurityScanAction',
                category: 'Test',
                owner: 'AWS',
                provider: 'CodeBuild',
                version: '1',
                inputArtifacts: ['BuildOutput'],
                outputArtifacts: ['SecurityOutput'],
                configuration: {
                  ProjectName: securityProjectName,
                },
              },
            ],
          },
          // Stage 5: Approval + Deploy
          {
            name: 'Deploy',
            actions: [
              {
                name: 'ManualApproval',
                category: 'Approval',
                owner: 'AWS',
                provider: 'Manual',
                version: '1',
                configuration: {
                  CustomData:
                    'Please review the build and security scan results before deploying.',
                },
                runOrder: 1,
              },
              {
                name: 'DeployAction',
                category: 'Invoke',
                owner: 'AWS',
                provider: 'Lambda',
                version: '1',
                inputArtifacts: ['BuildOutput'],
                configuration: {
                  FunctionName: deployFunctionName,
                },
                runOrder: 2,
              },
            ],
          },
        ],
        tags: {
          ...tags,
          Name: `cicd-pipeline-${environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create EventBridge rule to trigger pipeline on CodeCommit changes
    const triggerRule = new aws.cloudwatch.EventRule(
      `pipeline-trigger-rule-${environmentSuffix}`,
      {
        name: `cicd-pipeline-trigger-${environmentSuffix}`,
        description: 'Trigger pipeline on CodeCommit push to main branch',
        eventPattern: pulumi.interpolate`{
          "source": ["aws.codecommit"],
          "detail-type": ["CodeCommit Repository State Change"],
          "detail": {
            "event": ["referenceCreated", "referenceUpdated"],
            "repositoryName": ["${repositoryName}"],
            "referenceName": ["main"]
          }
        }`,
      },
      { parent: this }
    );

    const triggerRole = new aws.iam.Role(
      `pipeline-trigger-role-${environmentSuffix}`,
      {
        assumeRolePolicy: JSON.stringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Principal: { Service: 'events.amazonaws.com' },
              Action: 'sts:AssumeRole',
            },
          ],
        }),
      },
      { parent: this }
    );

    new aws.iam.RolePolicy(
      `pipeline-trigger-policy-${environmentSuffix}`,
      {
        role: triggerRole.id,
        policy: this.pipeline.arn.apply(arn =>
          JSON.stringify({
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: ['codepipeline:StartPipelineExecution'],
                Resource: arn,
              },
            ],
          })
        ),
      },
      { parent: this }
    );

    new aws.cloudwatch.EventTarget(
      `pipeline-trigger-target-${environmentSuffix}`,
      {
        rule: triggerRule.name,
        arn: this.pipeline.arn,
        roleArn: triggerRole.arn,
      },
      { parent: this }
    );

    this.registerOutputs({
      pipelineName: this.pipeline.name,
      pipelineArn: this.pipeline.arn,
    });
  }
}
